stacks-signer/src/signer.rs

```diff
@@ -1328,6 +1360,74 @@ impl Signer {
         Ok(())
     }
 
+    /// Load the saved signer state for the current dkg round
+    pub fn load_saved_state(&mut self) -> Result<(), PersistenceError> {
+        let Some(aggregate_key) = self.approved_aggregate_public_key else {
+            return Ok(());
+        };
+
+        info!("{self}: Loading saved state for key: {aggregate_key}");
+        if let Some(state) = self.load_saved_state_for_aggregate_key(aggregate_key)? {
+            let party_id = state.party_id;
+            let poly_commitment = state.get_poly_commitment(&mut OsRng);
+
+            let party_polynomials = poly_commitment
+                .as_ref()
+                .map(|poly_commitment| (&party_id, poly_commitment));
+
+            let mut dkg_results_bytes = vec![];
+
+            if let Err(e) = SignerMessage::serialize_dkg_result(
+                &mut dkg_results_bytes,
+                &aggregate_key,
+                party_polynomials.into_iter(),
+            ) {
+                error!(
+                    "{self}: Failed to serialize DKGResults message after loading saved state: {e}"
+                );
+            } else if let Err(e) = self
+                .stackerdb
+                .send_message_bytes_with_retry(&MessageSlotID::DkgResults, dkg_results_bytes)
+            {
```

Comment:
Why does this send a DkgResults message when it finishes loading saved state?

Reply:
This is to share the freshly loaded DKG shares with the Nakamoto sign coordinator. Without this the call to `get_signer_commitments` will load DKG shares from the last round which might not correspond to the one that was loaded by the signers.

I am not super happy with this solution, so if you have any improvement suggestions for how we can more properly communicate the appropriate shares to the sign coordinator, I'm all ears. The options I've considered were to add another message type for this message to more specifically convey that the signer has loaded a set of old shares, alternatively to implement persistence for old shares in the sign coordinator as well (which I'm not sure how complex it would be).

Reply:
Yes the miner-as-coordinator (`testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs`). It loads the DkgResults from stackerdb as you pointed out. The purpose of sending this message is to ensure that the latest DkgResults in stackerdb is the same as the ones the signer has just loaded on startup.

To clarify things consider the following scenario:
1. The signers run a successful DKG round and vote for the aggregate key for the current reward cycle.
2. For any reason, another DKG round was triggered and successfully run except that the voting in `signers-voting.clar` fails since an aggregate key is already set for the current reward cycle.

In this scenario, the latest DkgResults in stackerdb does not align with the approved aggregate key. Therefore, having the signers send new DkgResults messages whenever they load an old state ensures that the miner-as-coordinator can keep relying on fetching the latest DkgResults from stackerdb and still coordinate signing rounds.

Reply:
It seems like the issue then is step (2) of that scenario.

If its really not possible to make sure that signers don't try to assemble DkgResults messages for reward cycles that have already selected their aggregate keys, it may make more sense to use a new message for "DkgResult for approved aggregate key".

Reply:
Alright, based on your comments and our conversation yesterday I have updated the behavior so that the Signers will hold on to the DKG results message and send it only when an aggregate key is approved. This eliminates the need to also send this message when loading an old signer so I have removed that code.